### PR TITLE
Replace zeroMQ by tokio

### PR DESCRIPTION
**Pull request by marcusklaas**
This PR replaces the dependency on zeroMQ for message passing to and from the transactor. Since it uses the `tokio` crates, the communication is asynchronous in principle. But because the transactor is synchronized, it doesn't really add much. I guess the encoding/ decoding of objects is done concurrently at least.

The tests pass and it seems to work from my very light user testing. There's something weird going on with the `TxReports`, though. When adding new entities, the transactor doesn't report any new entity id's. It's especially weird since I didn't touch any of that (I think!). Does this behaviour sound familiar to you?

Oh and the reason the diff is so bloated is that we're using different versions of `rustfmt` I think. I'm on `0.9.0`. If you tell me yours, I'll format it using that version so the diff only highlight the actual changes.

### CODE REVIEW

**Comment by marcusklaas** [comment]
Hmmm, I seem to be on an old nightly locally. I'll update the PR.

**Comment by loganmhb** [comment]
Cool! I'll look this over. It looks like I'm on an older rustfmt (0.8.4) so I'll upgrade mine and push a formatting commit to master -- that should help the diff.

**Comment by loganmhb** [comment]
@marcusklaas hmm my formatting commit seems to have made things worse. But it looks good. If you can resolve the merge conflicts I'd be happy to merge it.

**Comment by marcusklaas** [comment]
Rebased!

File: src/backends/sqlite.rs [hunk 1]
```diff
@@ -9,11 +8,13 @@ use rmp_serde::{Serializer, Deserializer};
...
 use tx::TxRaw;
 
-#[derive(Clone)]
 pub struct SqliteStore {
-    conn: Arc<sql::Connection>,
+    conn: sql::Connection,
 }
 
+// FIXME: this is irresponsible!
+unsafe impl ::std::marker::Sync for SqliteStore {}
```
Comment (loganmhb) [inline src/backends/sqlite.rs]: Could we use an `Arc<Mutex<sql::Connection>>` in the `SqliteStore` instead to avoid this unsafety?

Comment (marcusklaas) [inline src/backends/sqlite.rs]: I think not. `Arc<Mutex<sql::Connection>>` will not `Sync` unless `sql::Connection` is `Sync`, right?

Comment (loganmhb) [inline src/backends/sqlite.rs]: Hmm, I guess that's right. I'll poke around at this a little bit.

Comment (marcusklaas) [inline src/backends/sqlite.rs]: Please do. Would be nice to get rid of this. For some more information, see [this issue](https://github.com/jgallagher/rusqlite/issues/188).

Comment (loganmhb) [inline src/backends/sqlite.rs]: Ah! Good news: `Mutex<T>` implements `Sync<T> where T: Send`, so wrapping the `SqliteStore`'s connection in a mutex looks like it does work. (Requires some minor changes to the trait implementation to lock the mutex, but that's all.)

My pull request workflow knowledge is failing me here, but I have a patch that you can apply in order to make the necessary change, if you like:
```
From e4cf01e253a4783dd03230df897d5979ab846471 Mon Sep 17 00:00:00 2001
From: Logan Buckley <logan.buckley@gmail.com>
Date: Wed, 12 Jul 2017 16:08:27 -0400
Subject: [PATCH] Make SqliteStore Sync via a mutex.

---
 src/backends/sqlite.rs | 29 ++++++++++++++---------------
 1 file changed, 14 insertions(+), 15 deletions(-)

diff --git a/src/backends/sqlite.rs b/src/backends/sqlite.rs
index 992adba..aa6e230 100644
--- a/src/backends/sqlite.rs
+++ b/src/backends/sqlite.rs
@@ -1,4 +1,5 @@
 use std::path::Path;
+use std::sync::{Arc, Mutex};
 
 use rusqlite as sql;
 
@@ -9,12 +10,9 @@ use {Result, KVStore, Record};
 use tx::TxRaw;
 
 pub struct SqliteStore {
-    conn: sql::Connection,
+    conn: Arc<Mutex<sql::Connection>>,
 }
 
-// FIXME: this is irresponsible!
-unsafe impl ::std::marker::Sync for SqliteStore {}
-
 impl SqliteStore {
     pub fn new<P: AsRef<Path>>(path: P) -> Result<SqliteStore> {
         let conn = sql::Connection::open(path)?;
@@ -30,7 +28,7 @@ impl SqliteStore {
         )?;
 
 
-        let store = SqliteStore { conn: conn };
+        let store = SqliteStore { conn: Arc::new(Mutex::new(conn)) };
 
         Ok(store)
     }
@@ -38,8 +36,8 @@ impl SqliteStore {
 
 impl KVStore for SqliteStore {
     fn get(&self, key: &str) -> Result<Vec<u8>> {
-        let mut stmt = self.conn
-            .prepare("SELECT val FROM logos_kvs WHERE key = ?1")
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare("SELECT val FROM logos_kvs WHERE key = ?1")
             .unwrap();
         stmt.query_row(&[&key], |row| {
             let s: Vec<u8> = row.get(0);
@@ -48,17 +46,18 @@ impl KVStore for SqliteStore {
     }
 
     fn set(&self, key: &str, value: &[u8]) -> Result<()> {
-        let mut stmt = self.conn
-            // We can't assume the key isn't already set, so need INSERT OR REPLACE.
-            .prepare("INSERT OR REPLACE INTO logos_kvs (key, val) VALUES (?1, ?2)")
-            .unwrap();
+        let conn = self.conn.lock().unwrap();
+        // We can't assume the key isn't already set, so need INSERT OR REPLACE.
+        let mut stmt = conn.prepare(
+            "INSERT OR REPLACE INTO logos_kvs (key, val) VALUES (?1, ?2)",
+        ).unwrap();
         stmt.execute(&[&key, &value])?;
         Ok(())
     }
 
     fn get_txs(&self, from: i64) -> Result<Vec<TxRaw>> {
-        let mut stmt = self.conn
-            .prepare("SELECT id, val FROM logos_txs WHERE id > ?1")
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare("SELECT id, val FROM logos_txs WHERE id > ?1")
             .unwrap();
         let results: Vec<Result<TxRaw>> = stmt.query_map(&[&from], |ref row| {
             let bytes: Vec<u8> = row.get(1);
@@ -85,8 +84,8 @@ impl KVStore for SqliteStore {
         let mut serialized: Vec<u8> = vec![];
         tx.records.serialize(&mut Serializer::new(&mut serialized))?;
 
-        let mut stmt = self.conn
-            .prepare("INSERT INTO logos_txs (id, val) VALUES (?1, ?2)")
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare("INSERT INTO logos_txs (id, val) VALUES (?1, ?2)")
             .unwrap();
 
         stmt.execute(&[&tx.id, &serialized])?;
-- 
2.13.2
```

Comment (marcusklaas) [inline src/backends/sqlite.rs]: Great! I've updated the pull request.

**Pull request merged**